Enable AKS autoscaler rule

ArmaanMcleod opened the pull request:
## PR Summary

Fix #218

Added a `Azure.AKS.AutoScaling` rule for release `GA` and baseline `2021_09`. 

The rule checks boths the following types:

- `Microsoft.ContainerService/managedClusters`
- `Microsoft.ContainerService/managedClusters/agentPools`

Also only checks agent pools of type `VirtualMachineScaleSets`, and agent pools that are any other type(such as `AvailabilitySets` just pass. Correct me if this needs to be handled differently. 

Also let me know if I should provide more detail in `docs/en/rules/Azure.AKS.AutoScaling.md`.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

ArmaanMcleod commented on docs/en/rules/Azure.AKS.AutoScaling.md:
```diff
@@ -0,0 +1,141 @@
+pillar: Performance Efficiency
+category: Performance
+resource: Azure Kubernetes Service
+online version: https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.AKS.AutoScaling/
```
Ah just realized this will be 404

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1:
```diff
@@ -138,6 +138,26 @@ Rule 'Azure.AKS.AzureRBAC' -Type 'Microsoft.ContainerService/managedClusters' -T
+    foreach ($agentPool in $agentPools) {
+
+        # Autoscaling only available on virtual machine scale sets
+        if ($Assert.HasFieldValue($agentPool, 'type', 'VirtualMachineScaleSets')) {
```
I think this line is not going to be performing how you intended.

`$Assert.HasFieldValue` is going to be returning an `AssertResult` which is always an object, and will also pass the null check on this line.
To get the result of `HasFieldValue`, use `$Assert.HasFieldValue($agentPool, 'type', 'VirtualMachineScaleSets').Result` or explictilty cast to a boolean. i.e. `[bool]$Assert.HasFieldValue($agentPool, 'type', 'VirtualMachineScaleSets')`. Personally I prefer the `.Result` option because it is clearer.

This is why `cluster-A`, and `cluster-B` in the unit tests are reporting as fail instead of pass.

The approach of only reporting on scale set node pools make sense to me.

ArmaanMcleod replied:
@BernieWhite Ah that makes sense, I had a look at the `AssertRule` class and didn't see that Result property: https://github.com/microsoft/PSRule/blob/main/src/PSRule/Runtime/AssertResult.cs#L33-L36

ArmaanMcleod replied:
Should be fixed now

BernieWhite commented on tests/PSRule.Rules.Azure.Tests/Azure.AKS.Tests.ps1:
```diff
@@ -230,6 +230,22 @@ Describe 'Azure.AKS' -Tag AKS {
+            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Fail' });
+            $ruleResult | Should -Not -BeNullOrEmpty;
+            $ruleResult | Should -HaveCount 4
+            $ruleResult.TargetName | Should -BeIn 'cluster-A', 'cluster-B', 'cluster-C', 'cluster-D';
```
In think cluster A, B should be returning Pass. Altnatively you could filter out clusters/ node pools that have no scale set node pools by using a `-If` pre-condition.

ArmaanMcleod replied:
Fixed

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1:
```diff
@@ -162,6 +182,7 @@ function global:GetAgentPoolProfiles {
                 type = $TargetObject.properties.type
                 maxPods = $TargetObject.properties.maxPods
                 orchestratorVersion = $TargetObject.properties.orchestratorVersion
+                enableAutoScaling = $TargetObject.properties.enableAutoScaling
```
I think this will also need to be inserted after line 175 `orchestratorVersion = $_.properties.orchestratorVersion` above.

Doesn't look like there is test coverage for this but the scenario where the agentPool is a nested sub-resource in the cluster i.e. In the `"resources": [ ]` of the cluster should ideally work as well.

ArmaanMcleod replied:
This was another thing I was unsure about. If you have sub resources that are invalid, the rule record still outcome shows up at `Pass`.

ArmaanMcleod replied:
Yeah so tested this with this template: 

```json
    "resources": [
        {
            "apiVersion": "2021-02-01",
            "dependsOn": [
                "[concat('Microsoft.Resources/deployments/', 'WorkspaceDeployment-20210814190241')]"
            ],
            "type": "Microsoft.ContainerService/managedClusters",
            "location": "[parameters('location')]",
            "name": "[parameters('resourceName')]",
            "properties": {
                "kubernetesVersion": "[parameters('kubernetesVersion')]",
                "enableRBAC": "[parameters('enableRBAC')]",
                "dnsPrefix": "[parameters('dnsPrefix')]",
                "agentPoolProfiles": [
                    {
                        "name": "agentpool1",
                        "osDiskSizeGB": "[parameters('osDiskSizeGB')]",
                        "count": 3,
                        "enableAutoScaling": true,
                        "minCount": 1,
                        "maxCount": 5,
                        "vmSize": "Standard_DS2_v2",
                        "osType": "Linux",
                        "storageProfile": "ManagedDisks",
                        "type": "VirtualMachineScaleSets",
                        "mode": "System",
                        "maxPods": 110,
                        "availabilityZones": null
                    }
                ],
                "networkProfile": {
                    "loadBalancerSku": "standard",
                    "networkPlugin": "[parameters('networkPlugin')]"
                },
                "apiServerAccessProfile": {
                    "enablePrivateCluster": "[parameters('enablePrivateCluster')]"
                },
                "addonProfiles": {
                    "httpApplicationRouting": {
                        "enabled": "[parameters('enableHttpApplicationRouting')]"
                    },
                    "azurepolicy": {
                        "enabled": "[parameters('enableAzurePolicy')]"
                    },
                    "omsAgent": {
                        "enabled": false
                    }
                }
            },
            "tags": {},
            "identity": {
                "type": "SystemAssigned"
            },
            "resources": [
                {
                    "type": "Microsoft.ContainerService/managedClusters/agentPools",
                    "apiVersion": "2021-02-01",
                    "name": "[concat(parameters('resourceName'), '/agentpool2')]",
                    "dependsOn": [
                        "[resourceId('Microsoft.ContainerService/managedClusters', parameters('resourceName'))]"
                    ],
                    "properties": {
                        "osDiskSizeGB": "[parameters('osDiskSizeGB')]",
                        "count": 3,
                        "enableAutoScaling": false,
                        "vmSize": "Standard_DS2_v2",
                        "osType": "Linux",
                        "storageProfile": "ManagedDisks",
                        "type": "VirtualMachineScaleSets",
                        "mode": "System",
                        "maxPods": 110,
                        "availabilityZones": null
                    }
                }
            ]
        }
    ]
```

Which gave the following:

```pwsh

    ____  _____ ____        __
   / __ \/ ___// __ \__  __/ /__
  / /_/ /\__ \/ /_/ / / / / / _ \
 / ____/___/ / _, _/ /_/ / /  __/
/_/    /____/_/ |_|\__,_/_/\___/

Using PSRule v1.6.0+473073bb8bfb8cc11051923545e290f916850110
Using PSRule.Rules.Azure v0.0.1

----------------------------
Explore documentation: https://microsoft.github.io/PSRule/
Contribute and find source: https://github.com/microsoft/PSRule
Report issues: https://github.com/microsoft/PSRule/issues
PSRule.Rules.Azure: https://aka.ms/ps-rule-azure
----------------------------

> akstestcluster : Microsoft.ContainerService/managedClusters [1/1]

   PASS  Azure.AKS.AutoScaling

Rules processed: 1, failed: 0, errored: 0
```

Seems like I need to check how these sub resources are being checked since this doesn't look right.

ArmaanMcleod replied:
I guess another challenge here is that I'm unable to debug this by adding breakpoints inside the rules. I usually add this VSCODE configuration to debug PowerShell interactively:

```json
{
    "name": "PowerShell: Interactive Session",
    "type": "PowerShell",
    "request": "launch",
    "cwd": ""
}
```

But doesn't seem to have any effect if I place breakpoints inside rules.

ArmaanMcleod replied:
I guess another issue I'm having is that I can't set breakpoints within the rules and step through with the VSCode debugger. I ussually debug these issues with Interactive sessions but for some reason when I run `Assert-PSRule` it doesn't pick them up.

BernieWhite replied:
Your example template looks correct to me.

I think the errors is related to line 173, `type = $_.type` which should be `type = $_.properties.type`.
This appears to be a bug with the existing code.

Because the node pools is not matching the scale set type it's being passed.

ArmaanMcleod replied:
Ah good catch, that went past me. Should be fixed now.

BernieWhite reviewed (changes requested):
@ArmaanMcleod Great work! A few suggestions but otherwise all good.

BernieWhite commented:
/azp run

BernieWhite commented:
/azp run

BernieWhite reviewed (commented):
@ArmaanMcleod Nice. Almost there. One minor tweak and all good.

ArmaanMcleod commented:
@BernieWhite Thanks for the help 🙂 . I fixed the typo in `GetAgentProfiles` and also added another cluster to test sub resources. Seems to be working fine now.

BernieWhite reviewed (approved):
@ArmaanMcleod All good.
